### ISSUE TITLE
cURL error 7: Failed to connect on flush

### ISSUE THREAD

**Issue by The-Don-Himself** [FriendsOfSymfony/FOSHttpCacheBundle#385]
I've been getting the occasional 500 error with the message 

````
[2017-09-20 01:27:41] app.CRITICAL: Request to caching proxy at 127.0.0.1:8080 failed with message "cURL error 7: Failed to connect to 127.0.0.1 port 80: Connection refused (see http://curl.haxx.se/libcurl/c/libcurl-errors.html)" {"exception":"[object] (FOS\\HttpCache\\Exception\\ProxyUnreachableException(code: 0): Request to caching proxy at 127.0.0.1:8080 failed with message \"cURL error 7: Failed to connect to 127.0.0.1 port 80: Connection refused (see http://curl.haxx.se/libcurl/c/libcurl-errors.html)\" at C:\\wamp64\\www\\campus-discounts.com\\vendor\\friendsofsymfony\\http-cache\\src\\Exception\\ProxyUnreachableException.php:35, Http\\Client\\Exception\\NetworkException(code: 0): cURL error 7: Failed to connect to 127.0.0.1 port 80: Connection refused (see http://curl.haxx.se/libcurl/c/libcurl-errors.html) at C:\\wamp64\\www\\campus-discounts.com\\vendor\\php-http\\guzzle6-adapter\\src\\Promise.php:121, GuzzleHttp\\Exception\\ConnectException(code: 0): cURL error 7: Failed to connect to 127.0.0.1 port 80: Connection refused (see http://curl.haxx.se/libcurl/c/libcurl-errors.html) at C:\\wamp64\\www\\campus-discounts.com\\vendor\\guzzlehttp\\guzzle\\src\\Handler\\CurlFactory.php:186)"} []
````

I found it strange because sometimes it works and sometimes it dint.

I finally found that it happens when I want to invalidate a large number of urls, easily reproducible like so

````
        $cacheManager = $this->get('fos_http_cache.cache_manager');
        for ($i = 1; $i <= 1000; $i++) {
            $cacheManager->invalidateRoute('index');
        }
        $cacheManager->flush();
````

I'm still digging around, but it seems this is an error with curl as the request never even reaches Varnish. Is there anyone experiencing the same with and/or has a workaround. Some type of batch flushing?

**Comment by The-Don-Himself** [comment]
Ok, so here is where things get a little weirder. If I execute the above command in batches of 100, when I finally reach 1000, it doesn't error out, but if I just come out with a large number say 1,500 it does, but again if incrementally executing it in batches until 1,500 it doesn't. Thoughts anyone?

**Comment by dbu** [comment]
hm. i am not too deep into curl async requests - maybe it sends too many requests in parallel with the large batch, in which case varnish might not be able to accept all of them fast enough. can you check if you can tweak how much async request curl sends out in parallel and reduce that number?

**Comment by dbu** [comment]
with your example, it also looks like a duplication elimination might be useful, btw ;-) but i think we should consider that an edge case and should rather ask people to write a decorator to do that around the CacheManager if they need that.

**Comment by The-Don-Himself** [comment]
I'm not entirely convinced that Varnish is the culprit here because I was monitoring it using varnishlog and it seems like the purge requests never came in the first place so as to be rejected, besides varnish normally handles thousands of concurrent requests per second although I've never benchmarked it (or seen one) against thousands of concurrent purge requests. I think the issue is with curl although I'm also not too deep into it. 

Anyway, I do think this is an edge case of 'sending too many requests at once'. For my own project I will queue each purge request in RabbitMQ and send one by one.

By the way, the example I gave wasn't really what I use but an example that I was able to reproduce the issue. My real issue occurred because I have a [social commerce site](https://campus-discounts.com) where discounts can be commented on, liked, shared, etc. I decided to embed that data in each discount in a MongoDB like fashion problem is a simple username change requires that I walk each discount fetching ids and send purge requests to Varnish, which is not a problem for normal users but super active users who've requires many requests. Anyway, for now the MQ solution would solve(postpone) my  problems but should this be pointed out somewhere? I don't mind closing the issue as an edge case.

**Comment by dbu** [comment]
it could also be a bug somewhere in curl with handling too many parallel requests. from what you describe, i think async through rabbitmq seems the best solution for your situation anyways. best tell rabbitmq just what data changed, and then have a worker that determines what needs to be invalidated, so that you don't need to do that work in the process that handles a web request...
